internal/settings/gandi.go

```diff
@@ -0,0 +1,147 @@
+	"github.com/qdm12/ddns-updater/internal/regex"
+)
+
+type gandi struct {
+	domain    string
+	host      string
+	ttl       int
+	ipVersion models.IPVersion
+	key       string
+}
+
+func NewGandi(data json.RawMessage, domain, host string, ipVersion models.IPVersion,
+	_ bool, matcher regex.Matcher) (s Settings, err error) {
+	extraSettings := struct {
+		Key string `json:"key"`
+		TTL int    `json:"ttl"`
+	}{}
+	if err := json.Unmarshal(data, &extraSettings); err != nil {
+		return nil, err
+	}
+	g := &gandi{
+		domain:    domain,
+		host:      host,
+		ipVersion: ipVersion,
+		key:       extraSettings.Key,
+		ttl:       extraSettings.TTL,
+	}
+	if err := g.isValid(); err != nil {
+		return nil, err
+	}
+	return g, nil
+}
+
+func (g *gandi) isValid() error {
+	if len(g.key) == 0 {
+		return ErrEmptyKey
+	}
+	return nil
+}
+
+func (g *gandi) String() string {
+	return toString(g.domain, g.host, constants.GANDI, g.ipVersion)
+}
+
+func (g *gandi) Domain() string {
+	return g.domain
+}
+
+func (g *gandi) Host() string {
+	return g.host
+}
+
+func (g *gandi) DNSLookup() bool {
+	return true
+}
+
+func (g *gandi) IPVersion() models.IPVersion {
+	return g.ipVersion
+}
+
+func (g *gandi) BuildDomainName() string {
+	return buildDomainName(g.host, g.domain)
+}
+
+func (g *gandi) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    models.HTML(fmt.Sprintf("<a href=\"http://%s\">%s</a>", g.BuildDomainName(), g.BuildDomainName())),
+		Host:      models.HTML(g.Host()),
+		Provider:  "<a href=\"https://www.gandi.net/\">gandi</a>",
+		IPVersion: models.HTML(g.ipVersion),
+	}
+}
+
+func (g *gandi) setHeaders(request *http.Request) {
+	setUserAgent(request)
+	setContentType(request, "application/json")
+	setAccept(request, "application/json")
+	request.Header.Set("X-Api-Key", g.key)
+}
+
+func (g *gandi) Update(ctx context.Context, client *http.Client, ip net.IP) (newIP net.IP, err error) {
+	recordType := A
+	var ipStr string
+	if ip.To4() == nil { // IPv6
+		recordType = AAAA
+		ipStr = ip.To16().String()
+	} else {
+		ipStr = ip.To4().String()
+	}
+
+	u := url.URL{
+		Scheme: "https",
+		Host:   "dns.api.gandi.net",
+		Path:   fmt.Sprintf("/api/v5/domains/%s/records/%s/%s", g.domain, g.host, recordType),
+	}
+
+	buffer := bytes.NewBuffer(nil)
+	encoder := json.NewEncoder(buffer)
+	requestData := struct {
+		Values [1]string `json:"rrset_values"`
+		TTL    int       `json:"rrset_ttl"`
+	}{
+		Values: [1]string{ipStr},
+		TTL: func() int {
+			ttl := 3600
+			if g.ttl != 0 {
+				ttl = g.ttl
+			}
+			return ttl
+		}(),
+	}
+	if err := encoder.Encode(requestData); err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrRequestEncode, err)
+	}
+
+	request, err := http.NewRequestWithContext(ctx, http.MethodPut, u.String(), buffer)
+	if err != nil {
+		return nil, err
+	}
+	g.setHeaders(request)
+
+	response, err := client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusCreated {
```

Comment:
Do we always get a status created, even if we update the record? 🤔

Reply:
yes, I agree that is weird but I didn't design this API 😆